Use workspace lints for crates in `compiler/`

nnethercote opened the pull request:
This is nicer and hopefully less error prone than specifying lints via bootstrap.

r? @jieyouxu

nnethercote committed:
Specify rust lints for `compiler/` crates via Cargo.

By naming them in `[workspace.lints.rust]` in the top-level
`Cargo.toml`, and then making all `compiler/` crates inherit them with
`[lints] workspace = true`. (I omitted `rustc_codegen_{cranelift,gcc}`,
because they're a bit different.)

The advantages of this over the current approach:
- It uses a standard Cargo feature, rather than special handling in
  bootstrap. So, easier to understand, and less likely to get
  accidentally broken in the future.
- It works for proc macro crates.

It's a shame it doesn't work for rustc-specific lints, as the comments
explain.

nnethercote committed:
Add `unreachable_pub` to the default lints for `compiler/` crates.

And fix the new errors in the handful of crates that didn't have a
`#![warn(unreachable_pub)]`.

nnethercote committed:
Remove `#![warn(unreachable_pub)]` from all `compiler/` crates.

(Except for `rustc_codegen_cranelift`.)

It's no longer necessary now that `unreachable_pub` is in the workspace
lints.

nnethercote committed:
Remove `[lints.rust]` section from `rustc_type_ir`.

It was added in #129523 to enable building on stable when there were
`cfg(bootstrap)` occurrences in the crate. But those are gone now, so
the section can be removed.

nnethercote commented on compiler/rustc_builtin_macros/Cargo.toml:
```diff
@@ -3,9 +3,9 @@ name = "rustc_builtin_macros"
-unexpected_cfgs = { level = "warn", check-cfg = ['cfg(llvm_enzyme)'] }
+# njn: is this needed?
+# [lints.rust]
+# unexpected_cfgs = { level = "warn", check-cfg = ['cfg(llvm_enzyme)'] }
```
I don't know if this needs to be here. There is a `cfg!(llvm_enzyme)` test within the crate, but I don't know how this lint would be triggered.

nnethercote replied:
I would like to, but if you inherit workspace lints with `[lints] workspace = true` you can't add extra lints in the `Cargo.toml`, see https://github.com/rust-lang/cargo/issues/13157. The suggested workaround is to instead use lint attributes in the code. Which I'm happy to do, but I don't know how to translate this lint (in particular the `check-cfg`) into a lint attribute. Maybe it's just a matter of adding `#[allow(unexpected_cfgs)]` on the single `cfg!(llvm_enzyme)` usage in the code? But I can't reproduce the warning either locally or on CI, so I can't test such a change. Help!

Urgau replied:
The cfg is already in the extra check-cfg list in bootstrap, I'm pretty sure you can just remove it as bootstrap already adds it, which I think is clearly visible by the fact that CI isn't complaining about missing it.

https://github.com/rust-lang/rust/blob/03eb45452305f2d52348279d0caa5fc1f12c438d/src/bootstrap/src/lib.rs#L80-L82

nnethercote replied:
Aha! That makes sense. Especially because there are some `cfg!(llvm_enzyme)` conditions in a few other crates that don't cause problems. Thank you.

petrochenkov commented:
Maybe add a tidy check making sure that every `compiler/*/Cargo.toml` file has `workspace = true`?

jieyouxu commented on src/bootstrap/src/core/builder/cargo.rs:
```diff
@@ -1072,12 +1072,17 @@ impl Builder<'_> {
+            // Cargo, these lints can be move to a `[workspace.lints.rustc]`
+            // section in the top-level `Cargo.toml`.
+            //
+            // NOTE: these flags are added to RUSTFLAGS, which is ignored when
+            // compiling proc macro crates such as `rustc_macros`,
+            // unfortunately.
```
Discussion: yeah this is unfortunate.

> One way to do this is by unconditionally passing an arbitrary flag and asserting its presence in each compiler crate. Ugly, but it could work I think.
>
> -- https://github.com/rust-lang/rust/pull/138035#issuecomment-2702886506

We should probably investigate if we want to switch over to *that* approach to make sure the lints get *consistently* applied to all compiler crates.

jieyouxu replied:
But for *this* PR, it's not a blocker. I find the workspace lints approach for non-internal lints still less "magical" than doing *all* of them here.

jieyouxu reviewed (approved):
This approach seems reasonable to me. I opened https://github.com/rust-lang/rust/issues/138106 to track an cleanup follow-up to handle them consistently, but is not a blocker for this PR.

EDIT: bootstrap thread: https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Consistent.20handling.20of.20compiler.2F*.20RUSTFLAGs.20and.20lints

jieyouxu commented:
Marking as S-waiting-on-team for discussion in https://rust-lang.zulipchat.com/#narrow/channel/326414-t-infra.2Fbootstrap/topic/Consistent.20handling.20of.20compiler.2F*.20RUSTFLAGs.20and.20lints.

nnethercote commented on Cargo.toml:
```diff
@@ -63,6 +63,18 @@ exclude = [
+#
+# NOTE: rustc-specific lints (e.g. `rustc::internal`) aren't supported by
+# Cargo. They are specified for `compiler/` crates in
+# `src/bootstrap/src/core/builder/cargo.rs`.
```
@ehuss said that #44690 is probably a blocker. I updated the comment to mention that issue.

nnethercote committed:
Remove `[lints.rust]` section from `rustc_builtin_macros`.

`llvm_enzyme` is now in the extra `check-cfg` list in bootstrap, so it
doesn't need to be handled explicitly here.

nnethercote commented:
@jieyouxu: I think this is ready to go.
- The `llvm_enzyme` issue is fixed. There's a new commit removing it from `compiler/rustc_builtin_macros/Cargo.toml`.
- The Zulip discussion hasn't surfaced any alternative suggestions.
- I agree with @petrochenkov that a tidy check would be good. It will be similar to the existing `edition.rs` check, and I'm happy to write it, but it won't happen until next week. If you're willing to accept that in a follow-up PR then we could get this PR in the queue now.

jieyouxu commented:
Yes, I think the tidy check isn't a blocker and can be done separately. I'll file an issue so we don't forgor.

@rustbot label: -S-waiting-on-team +S-waiting-on-review

jieyouxu commented on Cargo.toml:
```diff
@@ -63,6 +63,20 @@ exclude = [
   "src/tools/x",
 ]
 
+# These lints are applied to many crates in the workspace. In practice, this is
+# all crates under `compiler/`.
+#
+# NOTE: rustc-specific lints (e.g. `rustc::internal`) aren't supported by
+# Cargo. (Support for them is possibly blocked by #44690 (attributes for
+# tools).) Those lints are instead specified for `compiler/` crates in
+# `src/bootstrap/src/core/builder/cargo.rs`.
+[workspace.lints.rust]
+# FIXME(edition_2024): Change this to `-Wrust_2024_idioms` when all of the
+# individual lints are satisfied.
+keyword_idents_2024 = "warn"
+unreachable_pub = "warn"
+unsafe_op_in_unsafe_fn = "warn"
```
Question: one more thing -- should we document (even just a brief mention about this) how to adjust non-tool lints for all compiler crates somewhere in rustc-dev-guide? (And mention that tool lints will have to be adjusted in bootstrap instead.)

nnethercote replied:
Maybe? I don't know what section it would go in.

jieyouxu replied:
Since we'll probably want to do something more consistently eventually, this comment is probably good enough.

jieyouxu commented on compiler/rustc_lint_defs/src/lib.rs:
```diff
@@ -1,7 +1,3 @@
-// tidy-alphabetical-start
-#![warn(unreachable_pub)]
-// tidy-alphabetical-end
```
lol

jieyouxu reviewed (approved):
Thanks! This is a nice cleanup, and since we don't really have any alternative better approach *yet*, let's land this to allow consistently enforcing `unreachable_pub` and whatever other lints we may wish to enable for *all* compiler crates.

I have one question re. maybe mentioning the `Cargo.toml` workspace lint section in rustc-dev-guide vs the tool lints in bootstrap cargo, since they're now in two places and there is a "better place" for non-tool lints to go (in `Cargo.toml`).

jieyouxu commented on Cargo.toml:
```diff
@@ -63,6 +63,20 @@ exclude = [
+# Cargo. (Support for them is possibly blocked by #44690 (attributes for
+# tools).) Those lints are instead specified for `compiler/` crates in
+# `src/bootstrap/src/core/builder/cargo.rs`.
+[workspace.lints.rust]
```
Suggestion: actually, can you leave a quick FIXME pointed to #138195 about the tidy check too?

nnethercote replied:
I think #138195 is enough, we don't need a comment in the code as well. I will do the tidy check next week. If I add a comment here I'll just have to take it out again there.

jieyouxu replied:
Sure, that's fine.

jieyouxu commented:
@bors r+

bors commented:
:pushpin: Commit 8a3e03392ef12508a3ce50875594ddedd1164ed7 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 8a3e03392ef12508a3ce50875594ddedd1164ed7 -->
<!-- homu: {"type":"Approved","sha":"8a3e03392ef12508a3ce50875594ddedd1164ed7","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nnethercote commented:
This ended up being reverted in https://github.com/rust-lang/rust/pull/138306 for breaking miri and other external tools, and being superseded by #138331.
